fsw/cfe-core/src/sb/cfe_sbr.h

```diff
@@ -0,0 +1,236 @@
+ * and not mixed together.
+ *
+ * Using this holding structure prevents assignment directly into a different index
+ * or direct usage as numeric value.
+ */
+typedef struct
+{
+    CFE_SB_RouteId_Atom_t RouteId; /**< \brief Holding value, do not use directly in code */
+} CFE_SBR_RouteId_t;
+
+/** \brief Routing table entry */
+typedef struct
+{
+     CFE_SB_MsgId_t        MsgId;        /**< \brief Message ID associated with route */
+     uint16                Destinations; /**< \brief Destinations counter TODO delete */
+     uint32                SeqCnt;       /**< \brief Message sequence counter */
+     CFE_SB_DestinationD_t *ListHeadPtr; /**< \brief Destination list head */
+} CFE_SBR_RouteEntry_t;
+
+/** \brief Module data */
+typedef struct
+{
+    CFE_SBR_RouteId_t     MsgMap[CFE_SBR_MSG_MAP_SIZE];            /**< \brief Message map */
+    CFE_SBR_RouteEntry_t  RoutingTbl[CFE_PLATFORM_SB_MAX_MSG_IDS]; /**< \brief Routing table */
+    CFE_SB_RouteId_Atom_t RouteIdxTop;                             /**< \brief First unused entry in RoutingTbl */
+} cfe_sbr_t;
+
+/** \brief For each id callback function prototype */
+typedef void (*CFE_SBR_CallbackPtr_t)(CFE_SBR_RouteId_t RouteId, void *ArgPtr);
+
+/******************************************************************************
+ * Function prototypes
+ */
+
+/**
+ *  \brief Initialize software bus routing module
+ */
+void CFE_SBR_Init(void);
+
+/**
+ *  \brief Add a route for the given a message id
+ *
+ *  Called for the first subscription to a message ID, uses up one
+ *  element in the routing table.  Assumes check for existing
+ *  route was already performed or routes could leak
+ *
+ *  \param[in] MsgId Message ID of the route to add
+ *
+ *  \returns Route ID
+ */
+CFE_SBR_RouteId_t CFE_SBR_AddRoute(CFE_SB_MsgId_t MsgId);
+
+/**
+ *  \brief Obtain the route id given a message id
+ *
+ *  \param[in] MsgId Message ID of the route to get
+ *
+ *  \returns Route ID
+ */
+CFE_SBR_RouteId_t CFE_SBR_GetRouteId(CFE_SB_MsgId_t MsgId);
+
+/**
+ *  \brief Obtain the message id given a route id
+ *
+ *  \param[in] RouteId Route ID of the message id to get
+ *
+ *  \returns Message ID
+ */
+CFE_SB_MsgId_t CFE_SBR_GetMsgId(CFE_SBR_RouteId_t RouteId);
+
+/**
+ *  \brief Obtain the destination list head pointer given a route id
+ *
+ *  \param[in] RouteId Route ID
+ *
+ *  \returns Destination list head pointer for the given route id
+ */
+CFE_SB_DestinationD_t *CFE_SBR_GetDestListHeadPtr(CFE_SBR_RouteId_t RouteId);
+
+/**
+ * \brief Set the destination list head pointer for given route id
+ *
+ * \param[in] RouteId Route Id
+ * \param[in] DestPtr Destination list head pointer
+ */
+void CFE_SBR_SetDestListHeadPtr(CFE_SBR_RouteId_t RouteId, CFE_SB_DestinationD_t *DestPtr);
+
+/* TODO add description */
+void CFE_SBR_IncrementSequenceCounter(CFE_SBR_RouteId_t RouteId);
+uint32 CFE_SBR_GetSequenceCounter(CFE_SBR_RouteId_t RouteId);
```

Comment:
Should we typedef the sequence count, for consistency?
(could argue either way, hard to fathom needing more than 32 bits, but still...)

Reply:
Now using the cfe message sequence counter typedef.